### DISCUSSION TITLE
Java Stream API에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by kunsanglee** [2024-woowacourse-study/level-interview#59] [Java]
.

**Comment by alstn113** [comment]
Java Stream API는 자바 8에서 소개된 기능으로, 컬렉션 요소를 처리하는데 사용됩니다. Stream API는 함수형 프로그래밍 스타일을 지원하여 코드를 간결하게 작성할 수 있도록 도와줍니다.

1. 스트림(Stream)

   - 데이터 요소를 시퀀스로 나타내는 연속적인 요소의 모음입니다.
   - 컬렉션, 배열 등 다양한 데이터 소스로부터 생성할 수 있습니다.
   - 요소들의 연속적인 흐름이므로, 한 번 소비하면 재사용할 수 없습니다.

2. 중간 연산

   - 스트림에 대해서 연속적으로 수행할 수 있는 중간 작업입니다.
   - 중간 연산을 사용하여 스트림을 변환, 필터링, 정렬할 수 있습니다.
   - 예시로는 filter(), map(), sorted() 등이 있습니다.

3. 종단 연산

   - 스트림의 요소에 최종 작업을 수행하는 연산입니다.
   - 종단 연산을 호출하면 스트림이 소비되며, 결과를 생성하거나, 컬렉션에 수집하거나, 출력할 수 있습니다.
   - 예시로는 forEach(), collect(), reduce() 등이 있습니다.

4. 순차 및 병렬 처리

   - Stream API를 사용하면 스트림을 순차적으로 또는 병렬로 처리할 수 있습니다.
   - 병렬 처리를 위해 멀티스레딩을 활용하며, parallel() 메서드를 사용하여 스트림을 병렬 스트림을 변환할 수 있습니다.

**Comment by GIVEN53** [comment]
Java 8부터 지원하며 컬렉션 처리에 대한 기능을 제공한다. 데이터의 종류와 관계없이 같은 방식으로 데이터를 처리할 수 있어 재사용성이 높다.

### 특징
1. 원본 데이터를 변경하지 않는다.
2. 내부 반복으로 데이터를 처리한다.
3. 일회용이다.

### 스트림 동작 단계
1. 스트림 생성
    - `컬렉션.stream()`, `Arrays.stream(배열)`, `Stream.generate()`, `Stream.iterate()` 등
2. 중간 연산
    - 데이터를 가공하거나 필터링하며 또 다른 스트림을 반환한다.
    - 중간 연산이 실제로 처리되는 시점은 최종 연산이다. -> **lazy**
    - `filter()`, `map()`, `sorted()`, `limit()`, `peek()`, ...
3. 최종 연산
    - lazy된 중간 연산을 모두 수행하여 데이터를 모두 소모한 후 스트림을 닫는다.
    - `forEach()`, `findxxx()`, `xxxMatch()`, `count()`, `sum()`, `collect()`, ...

**Comment by le2sky** [comment]
https://velog.io/@qwerty1434/%EC%8A%A4%ED%8A%B8%EB%A6%BC%EC%9D%98-%EA%B2%8C%EC%9C%BC%EB%A6%84

**Comment by GIVEN53** [comment]
컬렉션 처리
특징
원본 데이터 변경x
내부 반복
일회용 -> 재사용 불가
생성
중간 연산 -> lazy
최종 연산
